src/internal/connector/graph/middleware.go

```diff
@@ -232,18 +274,33 @@ func (middleware RetryHandler) retryRequest(
 			err)
 	}
 
-	if respErr != nil {
-		return nil, Stack(ctx, respErr).With("retry_count", executionCount)
+	if initialErr != nil {
+		return nil, stackReq(ctx, req, nil, initialErr)
 	}
 
 	return resp, nil
 }
 
-func (middleware RetryHandler) isRetriableErrorCode(req *http.Request, code int) bool {
-	return code == http.StatusInternalServerError || code == http.StatusServiceUnavailable
+var retryableRespCodes = []int{
+	http.StatusInternalServerError,
+	http.StatusServiceUnavailable,
+	http.StatusBadGateway,
+	http.StatusGatewayTimeout,
 }
 
-func (middleware RetryHandler) isRetriableRequest(req *http.Request) bool {
+func (mw RetryMiddleware) isRetriableRespCode(ctx context.Context, resp *http.Response, code int) bool {
+	if slices.Contains(retryableRespCodes, code) {
+		return true
+	}
+
+	// not a status code, but the message itself might indicate a connectivity issue that
+	// can be retried independent of the status code.
+	return strings.Contains(
+		strings.ToLower(getRespDump(ctx, resp, true)),
+		strings.ToLower(string(IOErrDuringRead)))
```

Comment:
Probably an overkill, but we could save on `ToLower` calls by making `IOErrDuringRead` lower case.

Reply:
Per company culture, minor performance improvements such as this are kept outside of our scope of concern.  O(n) growth performance or unbounded conditions, sure.  But we try not to sweat the little stuff.  What's more important in cases like this is whether the behavior causes unwanted outcomes (ie: unintended matches).